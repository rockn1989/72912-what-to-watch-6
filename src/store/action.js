export const ActionType = {
  SET_ACTIVE_GENRE: `films/set_genre`,
  SHOW_MORE: `films/show_more`,
  SET_FILMS_COUNTER: `films/set_films_counter`,
  LOAD_FILMS: `films/load_films`,
  LOAD_FILM: `films/id/load_film`,
  AUTHORIZED: `user/authorized`,
  SET_USER_INFO: `user/set_user_info`,
  SEND_COMMENT: `user/send_comment`,
  REDIRECT_TO_ROUTE: `redirect_to_route`,
  SEND_FORM_DATA: `form/send_form_data`,
  SHOW_ERROR: `show_error`
};

export const setGenreAction = (genre) => ({
  type: ActionType.SET_ACTIVE_GENRE,
  payload: genre
});

export const loadFilmsAction = (films) => ({
  type: ActionType.LOAD_FILMS,
  payload: films
});

export const loadFilmAction = (film) => ({
  type: ActionType.LOAD_FILM,
  payload: film
});

export const showMoreAction = () => ({
  type: ActionType.SHOW_MORE
});

export const setFilmsCounterAction = (counters) => ({
  type: ActionType.SET_FILMS_COUNTER,
  payload: counters
});

export const authorizationAction = (status) => ({
  type: ActionType.AUTHORIZED,
  payload: status
});

export const setUserInfoAction = (data) => ({
  type: ActionType.SET_USER_INFO,
  payload: data
});

export const redirectAction = (url) => ({
  type: ActionType.REDIRECT_TO_ROUTE,
  payload: url
});

export const sendUserCommentAction = (data) => ({
  type: ActionType.SEND_COMMENT,
  payload: data
});

export const sendFormDataAction = (bool) => ({
  type: ActionType.SEND_FORM_DATA,
  payload: bool
});

export const showErrorAction = (bool) => ({
  type: ActionType.SHOW_ERROR,
  payload: bool
});

