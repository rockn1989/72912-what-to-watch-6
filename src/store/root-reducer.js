import {combineReducers} from 'redux';

import {filmsData} from './films-data/films-data';
import {filmData} from './film-data/film-data';
import {user} from './user/user';
import {formStatus} from './form-status/form-status';
import {errorStatus} from './error-status/error-status';

export const NameSpace = {
  FILMS_DATA: `FILMS_DATA`,
  FILM_DATA: `FILM_DATA`,
  USER: `USER`,
  FORM_STATUS: `FORM_STATUS`,
  ERROR_STATUS: `ERROR_STATUS`
};

export default combineReducers({
  [NameSpace.FILMS_DATA]: filmsData,
  [NameSpace.FILM_DATA]: filmData,
  [NameSpace.USER]: user,
  [NameSpace.FORM_STATUS]: formStatus,
  [NameSpace.ERROR_STATUS]: errorStatus,
});

// export default combineReducers({
//   filmsData,
//   genre,
//   user,
//   formStatus,
//   errorStatus,
// });
