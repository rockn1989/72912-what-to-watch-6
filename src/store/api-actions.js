import {ActionCreator} from "./action";

export const loadFilmsList = () => (dispatch, _getState, api) => {
  api.get(`/films`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilms(data));
      dispatch(ActionCreator.filterFilms());
    });
};

export const loadFilm = (id) => (dispatch, _getState, api) => {
  api.get(`/films/${id}`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilm(data));
    })
    .catch(() => {
      dispatch(ActionCreator.redirect(`/page-not-found`));
    });
};

export const login = () => (dispatch, _getState, api) => {
  api.get(`/login`)
    .then(({data, status}) => {
      if (status !== 401) {
        dispatch(ActionCreator.authorization(true));
        dispatch(ActionCreator.setUserInfo(data));
      }
    });
};

export const logout = () => (dispatch, _getState, api) => {
  api.get(`/logout`)
    .then(({status}) => {
      if (status === 200) {
        dispatch(ActionCreator.authorization(false));
      }
    });
};

export const sendLogin = ({email, password}) => (dispatch, _getState, api) => {
  api.post(`/login`, {
    email,
    password
  }).then(({data, status}) => {
    if (status === 200) {
      dispatch(ActionCreator.authorization(true));
      dispatch(ActionCreator.setUserInfo(data));
      dispatch(ActionCreator.redirect(`/`));
    } else {
      dispatch(ActionCreator.authorization(false));
    }
  });
};

export const sendComment = ({id, comment, rating}) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.sendFormData(false));
  api.post(`/comments/${id}`, {
    rating,
    comment
  }).then(() => {
    dispatch(ActionCreator.sendFormData(true));
    dispatch(ActionCreator.redirect(`/films/${id}`));
  })
  .catch(() => {
    dispatch(ActionCreator.showError(true));
  });
};
