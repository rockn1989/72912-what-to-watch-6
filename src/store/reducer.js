import {ActionType} from './action';
import {AUTH, NO_AUTH} from '../const';

const initialState = {
  genre: `All genres`,
  film: {},
  films: [],
  filtredFilmsList: [],
  filmsCounter: [],
  authorizationStatus: NO_AUTH,
  userInfo: {
    avatarUrl: `img/avatar.jpg`
  }
};

const reducer = (state = initialState, action) => {

  switch (action.type) {
    case ActionType.LOAD_FILMS:
      return {
        ...state,
        films: action.payload
      };

    case ActionType.LOAD_FILM:
      return {
        ...state,
        film: action.payload
      };

    case ActionType.SET_ACTIVE_GENRE:
      return {
        ...state,
        genre: action.payload
      };

    case ActionType.FILTER_FILMS:
      return {
        ...state,
        filtredFilmsList: state.genre === initialState.genre ? [...state.films] : state.films.filter((film) => film.genre === state.genre)
      };

    case ActionType.SHOW_MORE:
      const genreCounter = state.filmsCounter.filter(({name}) => name === state.genre);
      const idx = state.filmsCounter.indexOf(...genreCounter);

      const newEl = {
        name: state.genre,
        counter: genreCounter[0].counter + 1
      };

      const newFilmsCounter = [...state.filmsCounter.slice(0, idx), newEl, ...state.filmsCounter.slice(idx + 1)];
      return {
        ...state,
        filmsCounter: [...newFilmsCounter]
      };

    case ActionType.SET_FILMS_COUNTER:
      const counters = [];
      for (let name of action.payload) {
        counters.push({
          name,
          counter: 0
        });
      }

      return {
        ...state,
        filmsCounter: [...counters]
      };

    case ActionType.RESET_FILMS_COUNTER:
      const resetFilmsCounter = state.filmsCounter.map((counter) => {
        return {
          name: counter.name,
          counter: 0
        };
      });
      return {
        ...state,
        filmsCounter: [...resetFilmsCounter]
      };

    case ActionType.AUTHORIZED:
      return {
        ...state,
        authorizationStatus: action.payload ? AUTH : NO_AUTH
      };

    case ActionType.SET_USER_INFO:
      const {avatar_url: avatar, ...restData} = {...action.payload};
      return {
        ...state,
        userInfo: {...restData, avatarUrl: avatar}
      };
  }
  return state;
};

export {reducer};
