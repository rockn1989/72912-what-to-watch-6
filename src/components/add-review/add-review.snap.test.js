import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import AddReview from './add-review';
import configureStore from 'redux-mock-store';
import {NameSpace} from '../../store/root-reducer';
import * as redux from 'react-redux';

const mockStore = configureStore({});
const store = mockStore({
  [NameSpace.USER]: {
    authorizationStatus: false,
    userInfo: {
      avatarUrl: `img/avatar.jpg`
    }
  },
  [NameSpace.FILMS_DATA]: {
    genre: `All genres`,
    filmsCounter: [],
    films: []
  },
  [NameSpace.FILM_DATA]: {
    film: {id: 1, genre: `Comedy`, name: `test`}
  },
  [NameSpace.FORM_STATUS]: {
    formStatus: true
  },
  [NameSpace.ERROR_STATUS]: {
    error: false
  }
});

it(`Should AddReview render correctly`, () => {
  jest.spyOn(redux, `useSelector`);
  const history = createMemoryHistory();
  const film = {id: 1, genre: `Comedy`, name: `test`};
  const ava = `img/avatar.jpg`;

  const {container} = render(
      <redux.Provider store={store}>
        <Router history={history}>
          <AddReview
            formStatus={true}
            auth={true}
            avatar={ava}
            film={film}
            onSendUserComment={jest.fn()}
            error={false}
          />
        </Router>
      </redux.Provider>
  );
  expect(container).toMatchSnapshot();
});
