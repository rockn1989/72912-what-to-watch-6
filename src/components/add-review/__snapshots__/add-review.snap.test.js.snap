// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should AddReview render correctly 1`] = `
<div>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__header"
    >
      <div
        class="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header undefined"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <nav
          class="breadcrumbs"
        >
          <ul
            class="breadcrumbs__list"
          >
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
                href="/films/1"
              >
                test
              </a>
            </li>
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
              >
                Add review
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="user-block"
        >
          <a
            class="user-block__link"
            href="/login"
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        class="movie-card__poster movie-card__poster--small"
      >
        <img
          alt="The Grand Budapest Hotel poster"
          height="327"
          width="218"
        />
      </div>
    </div>
    <div
      class="add-review"
    >
      <form
        action="#"
        class="add-review__form"
      >
        <div
          class="rating"
        >
          <div
            class="rating__stars"
          >
            <input
              class="rating__input"
              id="star-1"
              name="rating"
              type="radio"
              value="1"
            />
            <label
              class="rating__label"
              for="star-1"
            >
              Rating 
              1
            </label>
            <input
              class="rating__input"
              id="star-2"
              name="rating"
              type="radio"
              value="2"
            />
            <label
              class="rating__label"
              for="star-2"
            >
              Rating 
              2
            </label>
            <input
              class="rating__input"
              id="star-3"
              name="rating"
              type="radio"
              value="3"
            />
            <label
              class="rating__label"
              for="star-3"
            >
              Rating 
              3
            </label>
            <input
              class="rating__input"
              id="star-4"
              name="rating"
              type="radio"
              value="4"
            />
            <label
              class="rating__label"
              for="star-4"
            >
              Rating 
              4
            </label>
            <input
              class="rating__input"
              id="star-5"
              name="rating"
              type="radio"
              value="5"
            />
            <label
              class="rating__label"
              for="star-5"
            >
              Rating 
              5
            </label>
            <input
              class="rating__input"
              id="star-6"
              name="rating"
              type="radio"
              value="6"
            />
            <label
              class="rating__label"
              for="star-6"
            >
              Rating 
              6
            </label>
            <input
              class="rating__input"
              id="star-7"
              name="rating"
              type="radio"
              value="7"
            />
            <label
              class="rating__label"
              for="star-7"
            >
              Rating 
              7
            </label>
            <input
              class="rating__input"
              id="star-8"
              name="rating"
              type="radio"
              value="8"
            />
            <label
              class="rating__label"
              for="star-8"
            >
              Rating 
              8
            </label>
            <input
              class="rating__input"
              id="star-9"
              name="rating"
              type="radio"
              value="9"
            />
            <label
              class="rating__label"
              for="star-9"
            >
              Rating 
              9
            </label>
            <input
              class="rating__input"
              id="star-10"
              name="rating"
              type="radio"
              value="10"
            />
            <label
              class="rating__label"
              for="star-10"
            >
              Rating 
              10
            </label>
          </div>
        </div>
        <div
          class="add-review__text"
        >
          <textarea
            class="add-review__textarea"
            id="review-text"
            name="review-text"
            placeholder="Review text"
          />
          <div
            class="add-review__submit"
          >
            <button
              class="add-review__btn"
              disabled=""
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </form>
    </div>
  </section>
</div>
`;
